changelogs/unreleased/gh-7797-untrusted-input-validation.md

```diff
@@ -1,3 +1,5 @@
 ## bugfix/core
 
 * Added boundary checking for getenv() return values (gh-7797).
+* Introduced `os.getenv_limited` helper, exposed `os.PATH_MAX` to lua and
+  started using them in internal code instead of `os.getenv` (gh-7797).
```

Comment:
Ideally we should also update `tarantoolctl.in`, but we should do it with a caution: `tarantoolctl` may be run on older `tarantool` version, see #5429. There are two ways to update:

* With bump of the tarantool-common dependency in `debian/control` (see 575203f1b722bbfd0c5410ec8ab4c6bb78655d89).
* Add a local wrapper into `tarantoolctl.in` around `os.getenv_safe()` with fallback to `os.getenv()` if the former is not present.

I like the second, it is simpler.

Reply:
Thanks for pointing this out! Done:
```diff
diff --git a/extra/dist/tarantoolctl.in b/extra/dist/tarantoolctl.in
index 5ff25b9f5..d0c5e5263 100755
--- a/extra/dist/tarantoolctl.in
+++ b/extra/dist/tarantoolctl.in
@@ -25,6 +25,17 @@ int getppid(void);
 
 local TIMEOUT_INFINITY = 100 * 365 * 86400
 
+-- Use getenv_safe when it's present. When not, fall back to plain getenv.
+-- getenv_safe accepts name and size, getenv accepts only name. So we can call
+-- both functions with same arguments. Plain getenv will simply ignore the extra
+-- argument.
+local getenv = os.getenv_safe or os.getenv
+-- How many digits there are in a 8 byte integer + 1 digit for null terminator.
+-- UIDs and GIDs actually use 4 byte integers, but better be on the safe side.
+local UINT64_STRSIZE_MAX = 21
+local PATH_MAX = os.PATH_MAX or 1024
+local USER_SIZE_MAX = 257
+
 -- path of tarantoolctl binary
 local command_path = arg[0]
 -- name of tarantoolctl binary
@@ -64,16 +75,16 @@ local function shift_argv(arg, argno, argcount)
 end
 
 local function check_user_level()
-    local uid = os.getenv('UID')
-    if uid == 0 or os.getenv("NOTIFY_SOCKET") then
+    local uid = getenv('UID', UINT64_STRSIZE_MAX)
+    if uid == 0 or getenv("NOTIFY_SOCKET", PATH_MAX) then
         return nil
     end
     -- local dir configuration
-    local pwd = os.getenv('PWD')
+    local pwd = getenv('PWD', PATH_MAX)
     local udir = pwd and pwd .. '/.tarantoolctl'
     udir = udir and fio.stat(udir) and udir or nil
     -- or home dir configuration
-    local homedir = os.getenv('HOME')
+    local homedir = getenv('HOME', PATH_MAX)
     udir = udir or homedir and homedir .. '/.config/tarantool/tarantool'
     udir = udir and fio.stat(udir) and udir or nil
     -- if one of previous is not nil
@@ -476,12 +487,12 @@ local function wrapper_cfg(cfg)
     end
     -- force these startup options
     cfg.pid_file = default_cfg.pid_file
-    if os.getenv('USER') ~= default_cfg.username then
+    if getenv('USER', USER_SIZE_MAX) ~= default_cfg.username then
         cfg.username = default_cfg.username
     else
         cfg.username = nil
     end
-    if os.getenv("NOTIFY_SOCKET") then
+    if getenv("NOTIFY_SOCKET", PATH_MAX) then
         cfg.background = false
     elseif cfg.background == nil then
         cfg.background = true
```